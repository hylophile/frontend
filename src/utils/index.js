import dayjs from 'dayjs';
import React, { createElement } from 'react';
import { MemoryRouter } from 'react-router-dom';

import { getApplication, getAuthorization } from 'selectors/application';

/* eslint-disable react/display-name, react/prop-types */

/**
 * Creates an object containing action constants namespaced under the specified reducer.
 *
 * @param {string} reducer The name of the reducer or module
 * @param {string[]} actions A list of action names
 * @return {object} Object with action name keys and full action string values
 */
export function buildActions(reducer, actions) {
  const result = {};

  for (const action of actions) {
    result[action] = `${reducer}/${action}`;
  }

  return result;
}

/**
 * Creates a mock component which takes a set of props specified at runtime.
 *
 * @param {string} name The component/element name (in PascalCase)
 * @param {object} props An object containing props to setup
 * @return {object} Mock component with specified name and props
 */
export const mockComponent = (name, props = {}) => () =>
  createElement(name, props, props.children);

/**
 * Creates a mock component which will expose its props for snapshot testing purposes
 * @param {*} name The component/element name (in PascalCase)
 */
export const mockComponentWithProps = name => props =>
  createElement(name, props, props.children);

/**
 * Wraps a React component in a <MemoryRouter> suitable for testing
 *
 * @param {Component} WrappedComponent React component to wrap
 */
export const withMemoryRouter = (
  WrappedComponent,
  routerProps = {}
) => props => (
  <MemoryRouter {...routerProps}>
    <WrappedComponent {...props} />
  </MemoryRouter>
);

/**
 * Wraps a React component in a <ThemeWrapper> suitable for testing
 *
 * @param {Component} WrappedComponent React component to wrap
 */
export const withThemeProp = WrappedComponent => props => (
  <WrappedComponent theme={'default'} {...props} />
);

/**
 * Interpolate strings from the `params` object into the `url`.
 *
 * @param {object} endpoint Object with `url` and optional `params` object
 */
export const buildUrl = endpoint => {
  // this is defined by webpack.DefinePlugin
  const baseUrl = API_URL;
  const { url, params } = endpoint;

  if (!params) {
    return `${baseUrl}${url}`;
  }

  return Object.entries(params).reduce((resultUrl, entry) => {
    const [key, value] = entry;

    return resultUrl.replace(`{${key}}`, value);
  }, `${baseUrl}${url}`);
};

export const getInitialState = () => {
  // this needs to be require()d because an import results in a
  // circular dependency
  const { initialState } = require('reducers');

  try {
    const [accessToken, expiration] = [
      localStorage.getItem('accessToken'),
      localStorage.getItem('expiration')
    ];

    if (!accessToken) {
      return initialState;
    }

    const expirationDate = dayjs(JSON.parse(expiration));

    if (!expirationDate.isValid() || expirationDate.isBefore(dayjs())) {
      localStorage.removeItem('accessToken');
      localStorage.removeItem('expiration');
      return initialState;
    }

    return {
      ...initialState,
      application: {
        ...getApplication(initialState),
        authorization: {
          ...getAuthorization(initialState),
          accessToken: JSON.parse(accessToken),
          expiration: expirationDate
        }
      }
    };
  } catch {
    return initialState;
  }
};
