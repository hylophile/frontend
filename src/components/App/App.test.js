import dayjs from 'dayjs';
import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import ConnectedApp, { App } from './App';
import { withMemoryRouter, withThemeProp } from 'utils';
import { initialState } from 'reducers/application';

jest.mock('components/Footer/Footer', () =>
  require('utils').mockComponent('Footer')
);
jest.mock('components/Header/Header', () =>
  require('utils').mockComponent('Header')
);
jest.mock('components/ToastDrawer/ToastDrawer', () =>
  require('utils').mockComponent('ToastDrawer')
);
jest.mock('../ThemeHOC/withTheme', () => () => Component => props => (
  <Component {...props} />
));
jest.mock('pages', () => {
  const { mockComponent } = require('utils');

  return {
    Dashboard: mockComponent('Dashboard'),
    Favorites: mockComponent('Favorites'),
    Flavors: mockComponent('Flavors'),
    FlavorStash: mockComponent('FlavorStash'),
    Home: mockComponent('Home'),
    Login: mockComponent('Login'),
    NotFound: mockComponent('NotFound'),
    Profile: mockComponent('Profile'),
    Recipe: mockComponent('Recipe'),
    RecipeEditor: mockComponent('RecipeEditor'),
    Recipes: mockComponent('Recipes'),
    Register: mockComponent('Register'),
    ShoppingList: mockComponent('ShoppingList'),
    UserRecipes: mockComponent('UserRecipes'),
    UserSettings: mockComponent('UserSettings')
  };
});

describe('<App />', () => {
  const actions = {
    initApp: jest.fn()
  };
  const mockStore = configureStore();
  const store = mockStore({ application: initialState });
  const authedStore = mockStore({
    application: {
      ...initialState,
      authorization: {
        accessToken: '1234',
        expiration: dayjs().add(60, 'minutes')
      },
      user: {
        id: 123,
        name: 'Doe'
      },
      theme: {
        name: 'default'
      }
    }
  });

  afterEach(() => {
    actions.initApp.mockReset();
  });

  it('calls initApp', () => {
    const RoutedApp = withMemoryRouter(App);

    renderer.create(
      <Provider store={store}>
        <RoutedApp actions={actions} />
      </Provider>
    );

    expect(actions.initApp).toHaveBeenCalled();
  });

  it('renders home by default', () => {
    const RoutedApp = withMemoryRouter(ConnectedApp);
    const component = renderer.create(
      <Provider store={store}>
        <RoutedApp />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders login page', () => {
    const RoutedApp = withMemoryRouter(ConnectedApp, {
      initialEntries: ['/login']
    });
    const component = renderer.create(
      <Provider store={store}>
        <RoutedApp />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders register page', () => {
    const RoutedApp = withMemoryRouter(ConnectedApp, {
      initialEntries: ['/register']
    });
    const component = renderer.create(
      <Provider store={store}>
        <RoutedApp />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders profile page', () => {
    const RoutedApp = withMemoryRouter(ConnectedApp, {
      initialEntries: ['/user/profile']
    });
    const component = renderer.create(
      <Provider store={authedStore}>
        <RoutedApp />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders user recipes page', () => {
    const RoutedApp = withMemoryRouter(ConnectedApp, {
      initialEntries: ['/user/recipes']
    });
    const component = renderer.create(
      <Provider store={authedStore}>
        <RoutedApp />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders favorites page', () => {
    const RoutedApp = withMemoryRouter(ConnectedApp, {
      initialEntries: ['/user/favorites']
    });
    const component = renderer.create(
      <Provider store={authedStore}>
        <RoutedApp />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders flavor stash page', () => {
    const RoutedApp = withMemoryRouter(ConnectedApp, {
      initialEntries: ['/user/flavor-stash']
    });
    const component = renderer.create(
      <Provider store={authedStore}>
        <RoutedApp />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders user settings page', () => {
    const RoutedApp = withMemoryRouter(ConnectedApp, {
      initialEntries: ['/user/settings']
    });
    const component = renderer.create(
      <Provider store={authedStore}>
        <RoutedApp />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders shopping list page', () => {
    const RoutedApp = withMemoryRouter(ConnectedApp, {
      initialEntries: ['/user/shopping-list']
    });
    const component = renderer.create(
      <Provider store={authedStore}>
        <RoutedApp />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders recipe editor page', () => {
    const RoutedApp = withMemoryRouter(ConnectedApp, {
      initialEntries: ['/recipe/editor']
    });
    const component = renderer.create(
      <Provider store={authedStore}>
        <RoutedApp />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders dashboard page', () => {
    const RoutedApp = withMemoryRouter(ConnectedApp, {
      initialEntries: ['/dashboard']
    });
    const component = renderer.create(
      <Provider store={authedStore}>
        <RoutedApp />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders flavors page', () => {
    const RoutedApp = withMemoryRouter(ConnectedApp, {
      initialEntries: ['/flavors']
    });
    const component = renderer.create(
      <Provider store={store}>
        <RoutedApp />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders recipe page', () => {
    const RoutedApp = withMemoryRouter(ConnectedApp, {
      initialEntries: ['/recipe']
    });
    const component = renderer.create(
      <Provider store={store}>
        <RoutedApp />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('redirects to login for private route', () => {
    const ThemedApp = withThemeProp(ConnectedApp);
    const RoutedApp = withMemoryRouter(ThemedApp, {
      initialEntries: ['/user/recipes']
    });
    const component = renderer.create(
      <Provider store={store}>
        <RoutedApp />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
