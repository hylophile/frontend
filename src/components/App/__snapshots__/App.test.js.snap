// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> redirects to login for private route 1`] = `
Array [
  <Header />,
  <ToastDrawer />,
  <Login />,
  <Footer />,
]
`;

exports[`<App /> renders dashboard page 1`] = `
Array [
  <Header />,
  <ToastDrawer />,
  <Dashboard />,
  <Footer />,
]
`;

exports[`<App /> renders favorites page 1`] = `
Array [
  <Header />,
  <ToastDrawer />,
  <Favorites />,
  <Footer />,
]
`;

exports[`<App /> renders flavor stash page 1`] = `
Array [
  <Header />,
  <ToastDrawer />,
  <FlavorStash />,
  <Footer />,
]
`;

exports[`<App /> renders flavors page 1`] = `
Array [
  <Header />,
  <ToastDrawer />,
  <Flavors />,
  <Footer />,
]
`;

exports[`<App /> renders home by default 1`] = `
Array [
  <Header />,
  <ToastDrawer />,
  <Home />,
  <Footer />,
]
`;

exports[`<App /> renders login page 1`] = `
Array [
  <Header />,
  <ToastDrawer />,
  <Login />,
  <Footer />,
]
`;

exports[`<App /> renders profile page 1`] = `
Array [
  <Header />,
  <ToastDrawer />,
  <Profile />,
  <Footer />,
]
`;

exports[`<App /> renders recipe editor page 1`] = `
Array [
  <Header />,
  <ToastDrawer />,
  <RecipeEditor />,
  <Footer />,
]
`;

exports[`<App /> renders recipe page 1`] = `
Array [
  <Header />,
  <ToastDrawer />,
  <Recipe />,
  <Footer />,
]
`;

exports[`<App /> renders register page 1`] = `
Array [
  <Header />,
  <ToastDrawer />,
  <Register />,
  <Footer />,
]
`;

exports[`<App /> renders shopping list page 1`] = `
Array [
  <Header />,
  <ToastDrawer />,
  <ShoppingList />,
  <Footer />,
]
`;

exports[`<App /> renders user recipes page 1`] = `
Array [
  <Header />,
  <ToastDrawer />,
  <UserRecipes />,
  <Footer />,
]
`;

exports[`<App /> renders user settings page 1`] = `
Array [
  <Header />,
  <ToastDrawer />,
  <UserSettings />,
  <Footer />,
]
`;
